challenge_find_my_mates/src/find_my_mates.py

```diff
@@ -0,0 +1,338 @@
+                distance = entity.distance_to_2d(person.position)
+                if distance < distance_closest:
+                    temp_entity_near = entity
+                    distance_closest = distance
+            closest_entity.append(temp_entity_near)
+
+        self._robot.speech.speak('I found many people but none were as pretty as you.', block=True)
+        people_descriptions = {}
+        for i, person in enumerate(person_entities):
+            sentence = "I found {} near the {}.\n".format(getattr(person, 'name'), closest_entity[i].id)
+
+            unique_property = None
+            properties = [prop for prop in dir(person) if not prop.startswith('__')]
+            attributes = dict()
+            for prop in properties:
+                temp_prop = [getattr(person, prop) for person in person_entities]
+                c = Counter(temp_prop)
+                attributes[prop] = c[getattr(person, prop)]
+                if prop == 'gender_confidence':
+                    if any(j <= 68 for j in temp_prop):
+                        del attributes['gender']
+                    del attributes['gender_confidence']
+                del attributes['id']
+                for attr in attributes:
+                    if attributes[attr] == 1:
+                        unique_property = attr
+
+            # Edit these sentences based on available attributes
+            if unique_property == 'gender':
+                sentence += "and the person was a {}.".format(getattr(person, unique_property))
+            elif unique_property == 'tall':
+                sentence += "and the person was tall. "
+            elif unique_property == 'shirt_colour':
+                sentence += "and the person was wearing a mostly {} shirt.".format(getattr(person, unique_property))
+
+            people_descriptions[getattr(person, 'name')] = sentence
+            if i > 3:
+                break
+        n = 0
+        while n < 3:
+            self._robot.speech.speak('Which mate would you like me to describe?', block=True)
+            try:
+                result = self._robot.hmi.query('Name', 'T -> ' + ' | '.join(challenge_knowledge.common.names), 'T')
+                command_recognized = result.sentence
+            except TimeoutException:
+                command_recognized = None
+            if command_recognized == "":
+                self._robot.speech.speak("I am still waiting for a command and did not hear anything")
+            elif command_recognized in challenge_knowledge.common.names:
+                    self._robot.speech.speak("OK, I will describe {}".format(command_recognized), block=True)
+                    self._robot.speech.speak(people_descriptions[command_recognized], block=True)
+                    n += 1
+            else:
+                self._robot.speech.speak(
+                    "I don't understand, I expected a command like " + ", ".join(challenge_knowledge.names))
```

Comment:
A command... Names. But again the weird robot behaviour. Plus could remove the duplicate code?